include/zephyr/bluetooth/gatt.h

```diff
@@ -165,21 +165,117 @@ typedef ssize_t (*bt_gatt_attr_write_func_t)(struct bt_conn *conn,
 					     const void *buf, uint16_t len,
 					     uint16_t offset, uint8_t flags);
 
-/** @brief GATT Attribute structure. */
+/** @brief GATT Attribute
+ *
+ *  This type primarily represents an ATT Attribute that may be
+ *  an entry in the local ATT database. The objects of this type
+ *  must be part of an array that forms a GATT service.
+ *
+ *  While the formed GATT service is registered with the local
+ *  GATT server, pointers to this type can typically be given to
+ *  GATT server APIs, like bt_gatt_notify().
+ *
+ *  @note This type is given as an argument to the
+ *  bt_gatt_discover() application callback, but it's not a
+ *  proper object of this type. The permissions, read() and
+ *  write() methods are not available, and it's unsound to pass
+ *  the pointer to GATT server APIs.
+ */
 struct bt_gatt_attr {
-	/** Attribute UUID */
+	/** @brief Attribute Type, aka. "UUID"
+	 *
+	 *  The Attribute Type determines the interface can be expected
+	 *  from the read() and write() methods and the possible
+	 *  permission configurations.
+	 *
+	 *  E.g. Attribute of type @ref BT_UUID_GATT_CPF will act as a
+	 *  GATT Characteristic Presentation Format descriptor as
+	 *  specified in Core Specification 3.G.3.3.3.5.
+	 *
+	 *  Anyone can define a new Attribute Type.
+	 */
 	const struct bt_uuid *uuid;
-	/** Attribute read callback */
+
+	/** @brief Attribute Value read method
+	 *
+	 *  This is the generic interface to read from the Attribute
+	 *  Value of this object in on-air format.
+	 *
+	 *  If this is non-NULL, you may call it to read the value.
+	 *  Value is the Attribute Value
+	 *
+	 *  This function may safely assume the Attribute Permissions
+	 *  are satisfied for this read.
+	 *
+	 *  The read() and write() implementations can and often do have
+	 *  side effects besides reading and writing the value.
+	 *
+	 *  This function is primarily invoked by the GATT server when a
+	 *  remote GATT client performs a read, but it can also be
+	 *  invoked for a local read.
+	 *
+	 *  Must be NULL if the attribute is not readable.
+	 */
 	bt_gatt_attr_read_func_t read;
-	/** Attribute write callback */
+
+	/** @brief Attribute Value write method
+	 *
+	 *  This is the generic interface to write to the Attribute
+	 *  Value of this object in on-air format.
+	 *
+	 *  This function may safely assume the Attribute Permissions
+	 *  are satisfied for this write.
+	 *
+	 *  The read() and write() implementations can and often do have
+	 *  side effects besides reading and writing the value.
+	 *
+	 *  This function is primarily invoked by the GATT server when a
+	 *  remote GATT client performs a write, but it can also be
+	 *  invoked for a local write.
+	 *
+	 *  Must be NULL if the attribute is not writable.
+	 */
 	bt_gatt_attr_write_func_t write;
-	/** Attribute user data */
+
+	/** @brief User data for the read() and write()
+	 *
+	 *  The meaning of this field varies and is not specified here.
```

Comment:
Is it overkill to mention that the user data from e.g. `BT_GATT_CHARACTERISTIC` is the value here? 

Basically we have 3 types of user data:
1) User data supplied by the application with e.g. `BT_GATT_CHARACTERISTIC`
2) User data controlled by `gatt.c` with e.g. `BT_GATT_PRIMARY_SERVICE` (where the service is the user data)
3) User data controlled by `gatt.c` in the discovery callbacks

The first type of user data is quite useful, but we can only really rely on that when the UUID is `BT_UUID_GATT_CHRC`

Reply:
> The first type of user data is quite useful, but we can only really rely on that when the UUID is `BT_UUID_GATT_CHRC`

Pitt fall here. The attribute type alone does not determine the user data type. The application could in principle make its own implementation for `BT_UUID_GATT_CHRC`.

Reply:
> Is it overkill to mention that the user data from e.g. `BT_GATT_CHARACTERISTIC` is the value here?

I'm not against adding, but I'm not sure what to write. I think it may be cleaner that the macro doc explains what's happening. `BT_GATT_CHARACTERISTIC` creates more than one attribute, so it's kind of complicated. I'm not going to add anything here, but if you have an idea, let's discuss it as a separate PR.

Reply:
> > The first type of user data is quite useful, but we can only really rely on that when the UUID is `BT_UUID_GATT_CHRC`
> 
> Pitt fall here. The attribute type alone does not determine the user data type. The application could in principle make its own implementation for `BT_UUID_GATT_CHRC`.

Not sure I follow. `BT_UUID_GATT_CHRC` is just the UUID, right? And it cannot use the GATT API without `bt_gatt_attr`, so in what case is the `user_data` of a `BT_UUID_GATT_CHRC` not a value defined by the application?

Reply:
The application could make a characteristic declaration attribute that does not have `bt_gatt_attr_read_chrc` as its `read()` method, but something custom instead, which uses `user_data` in some incompatible. The custom `read()` method still has behave according to spec, but `user_data` is an internal detail. Maybe it's in big-endian or something.

Reply:
Ah, right. You are correct that we cannot assume anything for `BT_UUID_GATT_CHRC`, and my 1. point only applied for the characteristic value attribute (which is initialized with the UUID provided by the application). 

Thanks for reminding me